### DISCUSSION TITLE
How to mock library?

### DISCUSSION THREAD

**Discussion by shawngustaw** [Lokicoule/nestjs-cognito#432] [Q&A]
Is it correct that `@nestjs-cognito/testing` package still hits a live cognito service? If so is there a better way to mock the library entirely? This might be something nice to add to the docs 👍

**Comment by Lokicoule** [comment]
Hey @shawngustaw,

Thanks for the question. Yes, you're correct. The `@nestjs-cognito/testing` package is not intended to mock AWS Cognito. Instead, the `CognitoTestingController` provided by the package is used for e2e testing.

The `CognitoTestingController` exposes a single endpoint, `/cognito-testing-login`, which uses the HTTP POST method. This endpoint allows you to test the authentication process with AWS Cognito by sending a request with the `username`, `password`, and `clientId` in the request body. It handles the scenario where the user is not activated by default, making it useful for testing and verifying the behavior of AWS Cognito. It serves as a reference point and can help identify if the Cognito User Pool is misconfigured, as it has already been tested and is known to be working in a properly configured environment.

If you're looking for instructions on setting up the testing environment using `@nestjs-cognito/testing`, you can find more details in the [wiki](https://github.com/Lokicoule/nestjs-cognito/wiki).

However, if you have a specific use case that requires mocking AWS Cognito, I guess you can easily achieve it using Jest.

Don't know if it's clear, but I hope that helps you to understand the goal behind this package ^^

**Reply by jpamdeth** [reply]
I am also having trouble mocking the dependencies injected by @nestjs-cognito/core.  My service class has this constructor:
```
  constructor(private prisma: PrismaService,
              @InjectCognitoIdentityProvider()
              private readonly client: CognitoIdentityProvider) {}
```

My test has this:
```
  let identityService: IdentityService;
  let prismaService: DeepMocked<PrismaService>;
  let client: DeepMocked<CognitoIdentityProvider>;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        IdentityService,
        {
          provide: PrismaService,
          useValue: createMock<PrismaService>(),
        },
        {
          provide: CognitoIdentityProvider,
          useValue: createMock<CognitoIdentityProvider>(),
        }
      ],
    }).compile();

    identityService = module.get<IdentityService>(IdentityService);
    prismaService = module.get(PrismaService);
    client = module.get(CognitoIdentityProvider);
  });
```
But I'm getting an error that the dependencies are not satisfied:
`    Nest can't resolve dependencies of the IdentityService (PrismaService, ?). Please make sure that the argument "COGNITO_IDENTITY_PROVIDER_INSTANCE_TOKEN" at index [1] is available in the RootTestModule context.`

Is it not possible to use the @InjectCognitoIdentityProvider() decorator and still inject the mock to satisfy the dependency?

**Reply by Lokicoule** [reply]
I guess you should replace the provided key with `COGNITO_IDENTITY_PROVIDER_INSTANCE_TOKEN` to make it works:

```ts
import { COGNITO_IDENTITY_PROVIDER_INSTANCE_TOKEN } from "@nestjs-cognito/core";

  let identityService: IdentityService;
  let prismaService: DeepMocked<PrismaService>;
  let client: DeepMocked<CognitoIdentityProvider>;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        IdentityService,
        {
          provide: PrismaService,
          useValue: createMock<PrismaService>(),
        },
        {
          provide: COGNITO_IDENTITY_PROVIDER_INSTANCE_TOKEN,
          useValue: createMock<CognitoIdentityProvider>(),
        }
      ],
    }).compile();

    identityService = module.get<IdentityService>(IdentityService);
    prismaService = module.get(PrismaService);
    client = module.get(CognitoIdentityProvider);
  });
```

**Reply by jpamdeth** [reply]
This version worked:
```
  let identityService: IdentityService;
  let prismaService: DeepMocked<PrismaService>;
  let client: DeepMocked<CognitoIdentityProvider>;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        IdentityService,
        {
          provide: PrismaService,
          useValue: createMock<PrismaService>(),
        },
        {
          provide: COGNITO_IDENTITY_PROVIDER_INSTANCE_TOKEN,
          useValue: createMock<CognitoIdentityProvider>(),
        }
      ],
    }).compile();

    identityService = module.get<IdentityService>(IdentityService);
    prismaService = module.get(PrismaService);
    client = module.get(COGNITO_IDENTITY_PROVIDER_INSTANCE_TOKEN);
  });
```
